File: api/app/modules/extension/validate.test.js
const detectField = require('./detect-field');
const { validate } = require('./validate');

jest.mock('../../../../database/config', () => ({
  speciesID: { 9060: 'Homo sapiens' },
}));
jest.mock('./detect-field');
detectField.mockReturnValue('gene');

describe('Validate query params', () => {
  it('should return error with invalid species', () => {
    const expected = {
      err: true,
      message: 'Invalid species',
    };
    expect(validate('Unknown', 'auto', 'test')).toEqual(expected);
  });

  it('should return error with invalid field', () => {
    const expected = {
      err: true,
      message: 'Invalid query field',
    };
    expect(validate('Homo sapiens', 'test', 'test')).toEqual(expected);
  });

  describe('term errors', () => {
    it('should return error with missing term', () => {
      const expected = {
        err: true,
        message: 'Query term must be a string or number',
      };
      expect(validate('Homo sapiens', 'auto', '')).toEqual(expected);
    });

    it('should return error term not string or number', () => {
      const expected = {
        err: true,
        message: 'Query term must be a string or number',
      };
      expect(validate('Homo sapiens', 'auto', [])).toEqual(expected);
    });
  });

  describe('field validation', () => {
    it('should autodetect field', () => {
      const expected = {
        field: 'gene',
      };
      expect(validate('Homo sapiens', 'auto', 'test').field).toBe(expected.field);
    });

    it('should validate accepted field', () => {
      const expected = {
        field: 'uniprot',
      };
      expect(validate('Homo sapiens', 'uniprot', 'test').field).toBe(expected.field);
    });
  });

  describe('term validation', () => {
    it('should convert geneid to number', () => {
      const expected = {
        term: 1234,
      };
      expect(validate('Homo sapiens', 'geneid', '1234').term).toBe(expected.term);
    });

    it('should convert nextprot to uniprot', () => {
      const expected = {
        field: 'uniprot',
        species: 'Homo sapiens',
        term: 'Q9BUL8',
      };
      expect(validate('Homo sapiens', 'nextprot', 'NX_Q9BUL8')).toEqual(expected);
    });

    it('should string trailing decimal from refseq', () => {
      const expected = {
        term: 'NP_009148',
      };
      expect(validate('Homo sapiens', 'refseq', 'NP_009148.2').term).toBe(expected.term);
    });

    it('should keep non geneid fields as strings', () => {
      const expected = {
        term: 'test',
      };
      expect(validate('Homo sapiens', 'uniprot', 'test').term).toBe(expected.term);
    });
  });

  it('should return valid query', () => {
    const expected = {
      field: 'uniprot',
      species: 'Homo sapiens',
      term: 'test',
    };
    expect(validate('Homo sapiens', 'uniprot', ' test ')).toEqual(expected);
  });
});

File: api-aws/lambda/internal/extension/validate_test.go
package extension

import (
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestValidateFields(t *testing.T) {
	type expected struct {
		fields Fields
		err    error
	}
	tests := map[string]struct {
		fields   Fields
		expected expected
	}{
		"missing identifier": {
			fields: Fields{identifier: "", species: "Homo sapiens", term: "ABC123"},
			expected: expected{
				fields: Fields{identifier: "", species: "Homo sapiens", term: "ABC123"},
				err:    errors.New("missing identifier"),
			},
		},
		"missing species": {
			fields: Fields{identifier: "gene", species: "", term: "ABC123"},
			expected: expected{
				fields: Fields{identifier: "gene", species: "", term: "ABC123"},
				err:    errors.New("missing species"),
			},
		},
		"missing term": {
			fields: Fields{identifier: "gene", species: "Homo sapiens", term: ""},
			expected: expected{
				fields: Fields{identifier: "gene", species: "Homo sapiens", term: ""},
				err:    errors.New("missing query term"),
			},
		},
		"valid fields": {
			fields: Fields{identifier: "gene", species: "Homo sapiens", term: "ABC123"},
			expected: expected{
				fields: Fields{identifier: "gene", species: "Homo sapiens", term: "ABC123"},
				err:    nil,
			},
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			fields, err := validateFields(test.fields)
			assert.Equal(t, test.expected.fields, fields)
			assert.Equal(t, test.expected.err, err)
		})
	}
}

func TestValidateIdentifier(t *testing.T) {
	type expected struct {
		value string
		err   error
	}
	tests := map[string]struct {
		identifier string
		expected   expected
	}{
		"missing identifier": {identifier: "", expected: expected{value: "", err: errors.New("missing identifier")}},
		"invalid identifier": {identifier: "unknown", expected: expected{value: "", err: errors.New("unrecognized identifier")}},
		"valid identifier":   {identifier: "gene", expected: expected{value: "gene", err: nil}},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			value, err := validateIdentifier(test.identifier)
			assert.Equal(t, test.expected.value, value)
			assert.Equal(t, test.expected.err, err)
		})
	}
}

func TestValidateSpecies(t *testing.T) {
	type expected struct {
		value string
		err   error
	}
	tests := map[string]struct {
		species  string
		expected expected
	}{
		"missing species": {species: "", expected: expected{value: "", err: errors.New("missing species")}},
		"invalid species": {species: "unknown", expected: expected{value: "", err: errors.New("unrecognized species")}},
		"valid species":   {species: "Homo sapiens", expected: expected{value: "Homo sapiens", err: nil}},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			value, err := validateSpecies(test.species)
			assert.Equal(t, test.expected.value, value)
			assert.Equal(t, test.expected.err, err)
		})
	}
}

func TestValidateTerm(t *testing.T) {
	type expected struct {
		value string
		err   error
	}
	tests := map[string]struct {
		term     string
		expected expected
	}{
		"missing term": {term: "", expected: expected{value: "", err: errors.New("missing query term")}},
		"valid term":   {term: "ABC123", expected: expected{value: "ABC123", err: nil}},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			value, err := validateTerm(test.term)
			assert.Equal(t, test.expected.value, value)
			assert.Equal(t, test.expected.err, err)
		})
	}
}
